src/lower/mod.rs

```diff
@@ -724,8 +720,19 @@ impl<'k> LowerGoal<Env<'k>> for Goal {
                 g.lower_quantified(env, ir::QuantifierKind::ForAll, ids),
             Goal::Exists(ref ids, ref g) =>
                 g.lower_quantified(env, ir::QuantifierKind::Exists, ids),
-            Goal::Implies(ref wc, ref g) =>
-                Ok(Box::new(ir::Goal::Implies(wc.lower(env)?, g.lower(env)?))),
+            Goal::Implies(ref wc, ref g, ref elaborate) => {
+                let mut where_clauses = wc.lower(env)?;
+                if *elaborate {
+                    where_clauses = ir::with_current_program(|program| {
+                        let program = program.expect("cannot elaborate without a program");
+                        where_clauses.into_iter()
+                                     .flat_map(|wc| wc.expanded(program))
+                                     .casted()
+                                     .collect()
+                    });
+                }
+                Ok(Box::new(ir::Goal::Implies(where_clauses, g.lower(env)?)))
+            }
```

Comment:
Really all we need is enough data to know how many parameters are declared on trait vs associated types. I think eventually we are going to have to refactor to support a more general "query-style" lowering, like rustc does, which should enable us to uncover *that* data without full access to other things. But, as you say, for now we don't support hypotheses in user-defined goals anyhow... (though it seems like we may want that eventually)